sdk/storage/Azure.Storage.Files.Shares/src/PartitionedDownloader.cs

```diff
@@ -0,0 +1,392 @@
+                long totalLength = ParseRangeTotalLength(initialResponse.Value.Details.ContentRange);
+                if (initialLength == totalLength)
+                {
+                    await CopyToAsync(
+                        initialResponse,
+                        destination,
+                        cancellationToken)
+                        .ConfigureAwait(false);
+                    return initialResponse.GetRawResponse();
+                }
+
+                // Capture the etag from the first segment and use it
+                // later to ensure the file doesn't change while we're
+                // downloading the remaining segments
+                ETag etag = initialResponse.GetRawResponse().Headers.ETag.GetValueOrDefault();
+
+                // Create a queue of tasks that will each download one segment
+                // of the file.  The queue maintains the order of the segments
+                // so we can keep appending to the end of the destination
+                // stream when each segment finishes.
+                var runningTasks = new Queue<Task<Response<ShareFileDownloadInfo>>>();
+                runningTasks.Enqueue(initialResponseTask);
+                if (_maxWorkerCount <= 1)
+                {
+                    // consume initial task immediately if _maxWorkerCount is 1 (or less to be safe). Otherwise loop below would have 2 concurrent tasks.
+                    await ConsumeQueuedTask().ConfigureAwait(false);
+                }
+
+                // Fill the queue with tasks to download each of the remaining
+                // ranges in the file
+                foreach (HttpRange httpRange in GetRanges(initialLength, totalLength))
+                {
+                    // Add the next Task (which will start the download but
+                    // return before it's completed downloading)
+                    runningTasks.Enqueue(_client.DownloadAsync(
+                        httpRange,
+                        rangeGetContentHash: false,
+                        conditions,
+                        cancellationToken));
+
+                    // If we have fewer tasks than alotted workers, then just
+                    // continue adding tasks until we have _maxWorkerCount
+                    // running in parallel
+                    if (runningTasks.Count < _maxWorkerCount)
+                    {
+                        continue;
+                    }
+
+                    // Once all the workers are busy, wait for the first
+                    // segment to finish downloading before we create more work
+                    await ConsumeQueuedTask().ConfigureAwait(false);
+                }
+
+                // Wait for all of the remaining segments to download
+                while (runningTasks.Count > 0)
+                {
+                    await ConsumeQueuedTask().ConfigureAwait(false);
+                }
```

Comment:
there's plenty of duplication from blobs. we should consider extracting common part from both and supply blob and share specific logic via injectable delegates. (see how uploader is implemented).